File: app/src/main/java/com/stoutner/privacybrowser/dialogs/SaveDialog.java
/*
 * Copyright © 2016-2020 Soren Stoutner <soren@stoutner.com>.
 *
 * This file is part of Privacy Browser <https://www.stoutner.com/privacy-browser>.
 *
 * Privacy Browser is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Privacy Browser is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Privacy Browser.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.stoutner.privacybrowser.dialogs;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.provider.DocumentsContract;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.DialogFragment;

import com.stoutner.privacybrowser.R;
import com.stoutner.privacybrowser.helpers.DownloadLocationHelper;

import java.io.File;

public class SaveDialog extends DialogFragment {
    // Declare the save listener.
    private SaveListener saveListener;

    // The public interface is used to send information back to the parent activity.
    public interface SaveListener {
        void onSave(int saveType, DialogFragment dialogFragment);
    }

    // Declare the class constants.
    public static final int SAVE_LOGCAT = 0;
    public static final int SAVE_ABOUT_VERSION_TEXT = 1;
    public static final int SAVE_ABOUT_VERSION_IMAGE = 2;
    private static final String SAVE_TYPE = "save_type";

    // Declare the class variables.
    String fileName;

    @Override
    public void onAttach(@NonNull Context context) {
        // Run the default commands.
        super.onAttach(context);

        // Get a handle for save listener from the launching context.
        saveListener = (SaveListener) context;
    }

    public static SaveDialog save(int saveType) {
        // Create an arguments bundle.
        Bundle argumentsBundle = new Bundle();

        // Store the arguments in the bundle.
        argumentsBundle.putInt(SAVE_TYPE, saveType);

        // Create a new instance of the save dialog.
        SaveDialog saveDialog = new SaveDialog();

        // Add the arguments bundle to the dialog.
        saveDialog.setArguments(argumentsBundle);

        // Return the new dialog.
        return saveDialog;
    }

    // `@SuppressLint("InflateParams")` removes the warning about using null as the parent view group when inflating the alert dialog.
    @SuppressLint("InflateParams")
    @Override
    @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        // Get a handle for the arguments.
        Bundle arguments = getArguments();

        // Remove the incorrect lint warning that the arguments might be null.
        assert arguments != null;

        // Get the arguments from the bundle.
        int saveType = arguments.getInt(SAVE_TYPE);

        // Get a handle for the activity and the context.
        Activity activity = requireActivity();
        Context context = requireContext();

        // Use an alert dialog builder to create the alert dialog.
        AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(context, R.style.PrivacyBrowserAlertDialog);

        // Get the current theme status.
        int currentThemeStatus = getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK;

        // Set the title and icon according to the type.
        switch (saveType) {
            case SAVE_LOGCAT:
                // Set the title.
                dialogBuilder.setTitle(R.string.save_logcat);

                // Set the icon according to the theme.
                if (currentThemeStatus == Configuration.UI_MODE_NIGHT_NO) {
                    dialogBuilder.setIcon(R.drawable.save_dialog_day);
                } else {
                    dialogBuilder.setIcon(R.drawable.save_dialog_night);
                }
                break;

            case SAVE_ABOUT_VERSION_TEXT:
                // Set the title.
                dialogBuilder.setTitle(R.string.save_text);

                // Set the icon according to the theme.
                if (currentThemeStatus == Configuration.UI_MODE_NIGHT_NO) {
                    dialogBuilder.setIcon(R.drawable.save_text_blue_day);
                } else {
                    dialogBuilder.setIcon(R.drawable.save_text_blue_night);
                }
                break;

            case SAVE_ABOUT_VERSION_IMAGE:
                // Set the title.
                dialogBuilder.setTitle(R.string.save_image);

                // Set the icon according to the theme.
                if (currentThemeStatus == Configuration.UI_MODE_NIGHT_NO) {
                    dialogBuilder.setIcon(R.drawable.images_enabled_day);
                } else {
                    dialogBuilder.setIcon(R.drawable.images_enabled_night);
                }
                break;
        }

        // Set the view.  The parent view is null because it will be assigned by the alert dialog.
        dialogBuilder.setView(activity.getLayoutInflater().inflate(R.layout.save_dialog, null));

        // Set the cancel button listener.  Using `null` as the listener closes the dialog without doing anything else.
        dialogBuilder.setNegativeButton(R.string.cancel, null);

        // Set the save button listener.
        dialogBuilder.setPositiveButton(R.string.save, (DialogInterface dialog, int which) -> {
            // Return the dialog fragment to the parent activity.
            saveListener.onSave(saveType, this);
        });

        // Create an alert dialog from the builder.
        AlertDialog alertDialog = dialogBuilder.create();

        // Remove the incorrect lint warning below that `getWindow()` might be null.
        assert alertDialog.getWindow() != null;

        // Get a handle for the shared preferences.
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getContext());

        // Get the screenshot preference.
        boolean allowScreenshots = sharedPreferences.getBoolean(getString(R.string.allow_screenshots_key), false);

        // Disable screenshots if not allowed.
        if (!allowScreenshots) {
            alertDialog.getWindow().addFlags(WindowManager.LayoutParams.FLAG_SECURE);
        }

        // The alert dialog must be shown before items in the layout can be modified.
        alertDialog.show();

        // Get handles for the layout items.
        EditText fileNameEditText = alertDialog.findViewById(R.id.file_name_edittext);
        Button browseButton = alertDialog.findViewById(R.id.browse_button);
        TextView fileExistsWarningTextView = alertDialog.findViewById(R.id.file_exists_warning_textview);
        TextView storagePermissionTextView = alertDialog.findViewById(R.id.storage_permission_textview);
        Button saveButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE);

        // Remove the incorrect lint warnings below that the views might be null.
        assert fileNameEditText != null;
        assert browseButton != null;
        assert fileExistsWarningTextView != null;
        assert storagePermissionTextView != null;

        // Update the status of the save button when the file name changes.
        fileNameEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // Do nothing.
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // Do nothing.
            }

            @Override
            public void afterTextChanged(Editable s) {
                // Get the current file name.
                String fileNameString = fileNameEditText.getText().toString();

                // Convert the file name string to a file.
                File file = new File(fileNameString);

                // Check to see if the file exists.
                if (file.exists()) {
                    // Show the file exists warning.
                    fileExistsWarningTextView.setVisibility(View.VISIBLE);
                } else {
                    // Hide the file exists warning.
                    fileExistsWarningTextView.setVisibility(View.GONE);
                }

                // Enable the save button if the file name is populated.
                saveButton.setEnabled(!fileNameString.isEmpty());
            }
        });

        // Set the file name according to the type.
        switch (saveType) {
            case SAVE_LOGCAT:
                // Use a file name ending in `.txt`.
                fileName = getString(R.string.privacy_browser_logcat_txt);
                break;

            case SAVE_ABOUT_VERSION_TEXT:
                // Use a file name ending in `.txt`.
                fileName = getString(R.string.privacy_browser_version_txt);
                break;

            case SAVE_ABOUT_VERSION_IMAGE:
                // Use a file name ending in `.png`.
                fileName = getString(R.string.privacy_browser_version_png);
                break;
        }

        // Instantiate the download location helper.
        DownloadLocationHelper downloadLocationHelper = new DownloadLocationHelper();

        // Get the default file path.
        String defaultFilePath = downloadLocationHelper.getDownloadLocation(context) + "/" + fileName;

        // Display the default file path.
        fileNameEditText.setText(defaultFilePath);

        // Hide the storage permission text view if the permission has already been granted.
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
            storagePermissionTextView.setVisibility(View.GONE);
        }

        // Handle clicks on the browse button.
        browseButton.setOnClickListener((View view) -> {
            // Create the file picker intent.
            Intent browseIntent = new Intent(Intent.ACTION_CREATE_DOCUMENT);

            // Set the intent MIME type to include all files so that everything is visible.
            browseIntent.setType("*/*");

            // Set the initial file name.
            browseIntent.putExtra(Intent.EXTRA_TITLE, fileName);

            // Set the initial directory if the minimum API >= 26.
            if (Build.VERSION.SDK_INT >= 26) {
                browseIntent.putExtra(DocumentsContract.EXTRA_INITIAL_URI, Environment.getExternalStorageDirectory());
            }

            // Request a file that can be opened.
            browseIntent.addCategory(Intent.CATEGORY_OPENABLE);

            // Launch the file picker.  There is only one `startActivityForResult()`, so the request code is simply set to 0, but it must be run under `activity` so the request code is correct.
            activity.startActivityForResult(browseIntent, 0);
        });

        // Return the alert dialog.
        return alertDialog;
    }
}

File: app/src/main/java/com/stoutner/privacybrowser/dialogs/SaveDialog.kt
/*
 * Copyright 2019-2022 Soren Stoutner <soren@stoutner.com>.
 *
 * This file is part of Privacy Browser Android <https://www.stoutner.com/privacy-browser-android>.
 *
 * Privacy Browser Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Privacy Browser Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Privacy Browser Android.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.stoutner.privacybrowser.dialogs

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.text.Editable
import android.text.InputType
import android.text.TextWatcher
import android.view.WindowManager
import android.widget.EditText
import android.widget.TextView

import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import androidx.preference.PreferenceManager

import com.stoutner.privacybrowser.R
import com.stoutner.privacybrowser.helpers.GetUrlSizeHelper

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

import java.net.URL

// Define the class constants.
private const val URL_STRING = "url_string"
private const val FILE_SIZE_STRING = "file_size_string"
private const val FILE_NAME_STRING = "file_name_string"
private const val USER_AGENT_STRING = "user_agent_string"
private const val COOKIES_ENABLED = "cookies_enabled"

class SaveDialog : DialogFragment() {
    // Declare the class variables.
    private lateinit var saveListener: SaveListener

    // The public interface is used to send information back to the parent activity.
    interface SaveListener {
        fun onSaveUrl(originalUrlString: String, fileNameString: String, dialogFragment: DialogFragment)
    }

    override fun onAttach(context: Context) {
        // Run the default commands.
        super.onAttach(context)

        // Get a handle for the save webpage listener from the launching context.
        saveListener = context as SaveListener
    }

    companion object {
        // `@JvmStatic` will no longer be required once all the code has transitioned to Kotlin.
        @JvmStatic
        fun saveUrl(urlString: String, fileSizeString: String, fileNameString: String, userAgentString: String, cookiesEnabled: Boolean): SaveDialog {
            // Create an arguments bundle.
            val argumentsBundle = Bundle()

            // Store the arguments in the bundle.
            argumentsBundle.putString(URL_STRING, urlString)
            argumentsBundle.putString(FILE_SIZE_STRING, fileSizeString)
            argumentsBundle.putString(FILE_NAME_STRING, fileNameString)
            argumentsBundle.putString(USER_AGENT_STRING, userAgentString)
            argumentsBundle.putBoolean(COOKIES_ENABLED, cookiesEnabled)

            // Create a new instance of the save webpage dialog.
            val saveDialog = SaveDialog()

            // Add the arguments bundle to the new dialog.
            saveDialog.arguments = argumentsBundle

            // Return the new dialog.
            return saveDialog
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        // Get the arguments from the bundle.
        val originalUrlString = requireArguments().getString(URL_STRING)!!
        val fileSizeString = requireArguments().getString(FILE_SIZE_STRING)!!
        val fileNameString = requireArguments().getString(FILE_NAME_STRING)!!
        val userAgentString = requireArguments().getString(USER_AGENT_STRING)!!
        val cookiesEnabled = requireArguments().getBoolean(COOKIES_ENABLED)

        // Use an alert dialog builder to create the alert dialog.
        val dialogBuilder = AlertDialog.Builder(requireContext(), R.style.PrivacyBrowserAlertDialog)

        // Set the title.
        dialogBuilder.setTitle(R.string.save_url)

        // Set the icon.
        dialogBuilder.setIcon(R.drawable.download)

        // Set the view.
        dialogBuilder.setView(R.layout.save_dialog)

        // Set the cancel button listener.  Using `null` as the listener closes the dialog without doing anything else.
        dialogBuilder.setNegativeButton(R.string.cancel, null)

        // Set the save button listener.
        dialogBuilder.setPositiveButton(R.string.save) { _: DialogInterface, _: Int ->
            // Return the dialog fragment to the parent activity.
            saveListener.onSaveUrl(originalUrlString, fileNameString, this)
        }

        // Create an alert dialog from the builder.
        val alertDialog = dialogBuilder.create()

        // Get a handle for the shared preferences.
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(requireContext())

        // Get the screenshot preference.
        val allowScreenshots = sharedPreferences.getBoolean(getString(R.string.allow_screenshots_key), false)

        // Disable screenshots if not allowed.
        if (!allowScreenshots) {
            alertDialog.window!!.addFlags(WindowManager.LayoutParams.FLAG_SECURE)
        }

        // The alert dialog must be shown before items in the layout can be modified.
        alertDialog.show()

        // Get handles for the layout items.
        val urlEditText = alertDialog.findViewById<EditText>(R.id.url_edittext)!!
        val fileSizeTextView = alertDialog.findViewById<TextView>(R.id.file_size_textview)!!
        val saveButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE)

        // Set the file size text view.
        fileSizeTextView.text = fileSizeString

        // Populate the URL edit text according to the type.  This must be done before the text change listener is created below so that the file size isn't requested again.
        if (originalUrlString.startsWith("data:")) {  // The URL contains the entire data of an image.
            // Get a substring of the data URL with the first 100 characters.  Otherwise, the user interface will freeze while trying to layout the edit text.
            val urlSubstring = originalUrlString.substring(0, 100) + "…"

            // Populate the URL edit text with the truncated URL.
            urlEditText.setText(urlSubstring)

            // Disable the editing of the URL edit text.
            urlEditText.inputType = InputType.TYPE_NULL
        } else {  // The URL contains a reference to the location of the data.
            // Populate the URL edit text with the full URL.
            urlEditText.setText(originalUrlString)
        }

        // Update the file size when the URL changes.
        urlEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                // Do nothing.
            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                // Do nothing.
            }

            override fun afterTextChanged(editable: Editable) {
                // Get the current URL to save.
                val urlToSave = urlEditText.text.toString()

                // Enable the save button if the URL is populated.
                saveButton.isEnabled = urlToSave.isNotEmpty()

                CoroutineScope(Dispatchers.Main).launch {
                    // Create a URL size string.
                    var urlSize: String

                    // Get the URL size on the IO thread.
                    withContext(Dispatchers.IO) {
                        // Get the URL size.
                        urlSize = GetUrlSizeHelper.getUrl(requireContext(), URL(urlToSave), userAgentString, cookiesEnabled)
                    }

                    // Display the updated URL.
                    fileSizeTextView.text = urlSize
                }
            }
        })

        // Return the alert dialog.
        return alertDialog
    }
}
